core/reflection.jvm/src/kotlin/reflect/jvm/internal/KCallableImpl.kt

```diff
@@ -112,65 +112,80 @@ internal abstract class KCallableImpl<out R> : KCallable<R>, KTypeParameterOwner
-                masks.add(mask)
-                mask = 0
+        val arguments = getAbsentArguments().apply {
+            if (isSuspend) {
+                // continuationArgument is tail of arguments
+                this[parameters.size] = continuationArgument
             }
+        }
+        val masks = IntArray((parameters.size + Integer.SIZE - 1) / Integer.SIZE)
 
+        var valueParameterIndex = 0
+        var anyOptional = false
+
+        for (parameter in parameters) {
             when {
                 args.containsKey(parameter) -> {
-                    arguments.add(args[parameter])
+                    arguments[parameter.index] = args[parameter]
                 }
+                // Absent value is already set at _absentArguments
                 parameter.isOptional -> {
-                    // For inline class types, the javaType refers to the underlying type of the inline class,
-                    // but we have to pass null in order to mark the argument as absent for InlineClassAwareCaller.
-                    arguments.add(if (parameter.type.isInlineClassType) null else defaultPrimitiveValue(parameter.type.javaType))
-                    mask = mask or (1 shl (index % Integer.SIZE))
+                    val maskIndex = valueParameterIndex / Integer.SIZE
+                    masks[maskIndex] = masks[maskIndex] or (1 shl (valueParameterIndex % Integer.SIZE))
                     anyOptional = true
                 }
-                parameter.isVararg -> {
-                    arguments.add(defaultEmptyArray(parameter.type))
-                }
+                parameter.isVararg -> {}
                 else -> {
                     throw IllegalArgumentException("No argument provided for a required parameter: $parameter")
                 }
             }
 
             if (parameter.kind == KParameter.Kind.VALUE) {
-                index++
+                valueParameterIndex++
             }
         }
 
-        if (continuationArgument != null) {
-            arguments.add(continuationArgument)
-        }
-
         if (!anyOptional) {
-            return call(*arguments.toTypedArray())
+            return call(*arguments)
         }
 
-        masks.add(mask)
-
         val caller = defaultCaller ?: throw KotlinReflectionInternalError("This callable does not support a default call: $descriptor")
 
-        arguments.addAll(masks)
-
-        // DefaultConstructorMarker or MethodHandle
-        arguments.add(null)
-
         @Suppress("UNCHECKED_CAST")
         return reflectionCall {
-            caller.call(arguments.toTypedArray()) as R
+            // +1 is argument for DefaultConstructorMarker or MethodHandle
+            val mergedArgs = arrayOfNulls<Any?>(arguments.size + masks.size + 1)
+            System.arraycopy(arguments, 0, mergedArgs, 0, arguments.size)
+            for (i in masks.indices) {
+                mergedArgs[i + arguments.size] = masks[i]
+            }
```

Comment:
Since it is not possible to copy from an `IntArray` to an `Array<Any?>` using `System.arraycopy`, a `for` loop is used.